### PR TITLE
API: preserve freq in DTI/TDI.factorize

### LINKED ISSUE #35563: BUG: freq set to NONE when resampling pd.Multiindex (introduced in v1.1.0)
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---
## Update/Summary

When dealing with a pd.multiindex the frequency `df.loc[i].index.freq` has not been set. This behavior has been introduced in v1.1.0

> ```
> idx = pd.Index(range(2), name="A")
> dti = pd.date_range("2020-01-01", periods=7, freq="D", name="B")
> mi = pd.MultiIndex.from_product([idx, dti])
> 
> df = pd.DataFrame(np.random.randn(14, 2), index=mi)
> 
> >>> df.loc[0].index
> ```
> 
> `df.loc[0].index` matches `dti`, so it would be nice to get the `freq="D"` back on that.

> using code sample in [#35563 (comment)](https://github.com/pandas-dev/pandas/issues/35563#issuecomment-670246757) points to #31315
> 
> [c988567](https://github.com/pandas-dev/pandas/commit/c9885673d6a6f9f433c570d65b664b0c2ec8c6f0) is the first bad commit
> commit [c988567](https://github.com/pandas-dev/pandas/commit/c9885673d6a6f9f433c570d65b664b0c2ec8c6f0)
> Author: jbrockmendel [jbrockmendel@gmail.com](mailto:jbrockmendel@gmail.com)
> Date: Sun Feb 9 07:01:20 2020 -0800
> 
> ```
> REF: tighten what we accept in TimedeltaIndex._simple_new (#31315)
> ```

#### Code Sample, a copy-pastable example

```python
import numpy as np
import pandas as pd

idx_level_0 = np.repeat([1, 2], 5)
dates = np.tile(
    ["2020-01-01", "2020-01-02", "2020-01-04", "2020-01-06", "2020-01-07"], 2
)
values1 = [1, 2, 3, 4, 5]
values2 = [6, 7, 8, 9, 10]

df = pd.DataFrame(
    {"idx_level_0": idx_level_0, "dates": dates, "values": [*values1, *values2]}
)
df["dates"] = pd.to_datetime(df["dates"])
df = df.set_index(["idx_level_0", "dates"], drop=True)

df = df.groupby("idx_level_0").resample("D", level="dates").last()

# The following assertion is working properly in pandas v1.0.5
# It throws an error in pandas v1.1.0
assert df.index.get_level_values(1).freq == "D"
```

#### Problem description

When resampling a `groupby`-object, the frequency will incorrectly be set to `None`.

#### Expected Output
The frequency should be set according to the resampled frequency.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : d9fff2792bf16178d4e450fe7384244e50635733
python           : 3.8.2.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19041
machine          : AMD64
processor        : Intel64 Family 6 Model 60 Stepping 3, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : de_DE.cp1252
pandas           : 1.1.0
numpy            : 1.18.4
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.1.1
setuptools       : 45.3.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.5.1
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.2.2
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : 1.5.0
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by jbrockmendel**
cc @simonjayhawkins 
closes #35563 

this is an old branch I resurrected and rebased, so has some excess cruft that would need to be removed if we want to move forward.

LMK if there are any tests you'd like to see added based on the other issue (or just go ahead and push them if you like)

### CODE REVIEW

**Comment by simonjayhawkins** [comment]
thanks @jbrockmendel I'll look in detail tomorrow and add tests as required.

File: pandas/tests/indexes/datetimes/test_datetime.py [hunk 1]
```diff
@@ -287,49 +290,65 @@ def test_factorize(self):
...
         exp_arr = np.array([0, 0, 1, 2, 0, 2], dtype=np.intp)
         exp_idx = DatetimeIndex(["2014-03", "2014-02", "2014-01"])
         arr, idx = idx2.factorize()
         tm.assert_numpy_array_equal(arr, exp_arr)
         tm.assert_index_equal(idx, exp_idx)
+        assert idx.freq == exp_idx.freq
 
-        # freq must be preserved
+    def test_factorize_preserves_freq(self):
+        # GH#33836 freq should be preserved
```
Comment (simonjayhawkins) [inline pandas/tests/indexes/datetimes/test_datetime.py:303]: need to change this issue number

File: pandas/tests/indexes/timedeltas/test_timedelta.py [hunk 2]
```diff
@@ -75,17 +75,26 @@ def test_factorize(self):
...
+        assert idx.freq == exp_idx.freq
 
         arr, idx = idx1.factorize(sort=True)
         tm.assert_numpy_array_equal(arr, exp_arr)
         tm.assert_index_equal(idx, exp_idx)
+        assert idx.freq == exp_idx.freq
 
-        # freq must be preserved
+    def test_factorize_preserves_freq(self):
+        # GH#33836 freq should be preserved
```
Comment (simonjayhawkins) [inline pandas/tests/indexes/timedeltas/test_timedelta.py:86]: and here

**Review by simonjayhawkins — COMMENTED** [review]
I assume we need to add special casing since extension array interface doesn't support the sort kwarg, xref #33838

**Comment by jreback** [comment]
this is going to be hard to backport

File: pandas/core/algorithms.py [hunk 3]
```diff
@@ -198,8 +200,16 @@ def _reconstruct_data(
     -------
     ExtensionArray or np.ndarray
     """
+    if isinstance(values, ABCExtensionArray) and values.dtype == dtype:
+        # Catch DatetimeArray/TimedeltaArray
+        return values
+
     if is_extension_array_dtype(dtype):
-        values = dtype.construct_array_type()._from_sequence(values)
+        cls = dtype.construct_array_type()
```
Comment (jreback) [inline pandas/core/algorithms.py:208]: elsif here (why are using isintance when we should be using is_extension_array_dtype)?

**Review by jreback — CHANGES_REQUESTED** [review]
if you can get this passing on the backport as well ok to merge, but pls put up a separate patch for that to see if its even possible.

**Comment by simonjayhawkins** [comment]
> if you can get this passing on the backport as well ok to merge, but pls put up a separate patch for that to see if its even possible.

sure

File: pandas/core/algorithms.py (hunk 3 above)
Comment (simonjayhawkins) [reply pandas/core/algorithms.py:209]: For the elif part of the comment, I think that Brock maybe keeping the early return separate for readability

Comment (jbrockmendel) [reply pandas/core/algorithms.py:209]: > elsif here

sure

> why are using isintance when we should be using is_extension_array_dtype

1) perf, is_foo_dtype is slow
2) to catch DTA/TDA

(... i think.  most of this is a resurrected branch from a ways back)

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
blocking as will need a release note before merge, i'll add if https://github.com/pandas-dev/pandas/pull/38137 successful

File: pandas/core/arrays/datetimelike.py [hunk 4]
```diff
@@ -1645,6 +1645,24 @@ def _with_freq(self, freq):
         arr._freq = freq
         return arr
 
+    # --------------------------------------------------------------
+
+    def factorize(self, na_sentinel=-1, sort: bool = False):
```
Comment (simonjayhawkins) [inline pandas/core/arrays/datetimelike.py:1650]: _values_for_factorize and _from_factorized in DatetimeLikeArrayMixin are now only used by PeriodArray? should these be moved?

**Comment by simonjayhawkins** [comment]
> this is going to be hard to backport

maybe doable, in https://github.com/pandas-dev/pandas/pull/38137, the ci has not completed yet but looks like tests are passing (except for np-dev which is unrelated and failing on other PRs)

File: pandas/core/arrays/datetimelike.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py:1650]: values_for_factorize is still used (incorrectly) in core.util.hashing, so for now this is still needed here

Comment (simonjayhawkins) [reply pandas/core/arrays/datetimelike.py:1650]: ok

**Comment by jreback** [comment]
yeah ok with just making this work here, can fixup later / cleanup

**Comment by simonjayhawkins** [comment]
@jreback it would be good to get this into master before 1.2.0rc0 if it's ready. not critical but would only need one backport. backporting to 1.1.5 does not block 1.2.0rc0

if more discussion needed, fine. no need to be under any pressure to merge.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@simonjayhawkins no i think this is fine. I am sure that @jbrockmendel will have a cleanup on top of this, but good for now. thanks.

**Pull request merged**